File: Semester 7/Computational Finance (MA 473)/MA473 Term Paper/Code/findUInitial.m
function u = findUInitial(b, h, K, nu, eta, r, sigma, lambda, option)
%This function finds initial value of u
%Input:
%   b       : |x| <= b
%   h       : Spatial step size
%   K       : Strike Price
%Output:
%   u       

M = round(2*b/h);
h = 2*b/M;

x = -b:h:b;
SbyK = exp(x);
if option == "call"
    v = max(0, K*SbyK-K);
else
    v = max(0, K-K*SbyK);
end

kappa = findKappa(nu, eta);
alpha = findAlpha(r, lambda, sigma, kappa);

u = (v.*exp(-alpha*x))';
end

File: Semester 7/Computational Finance (MA 473)/Term Paper - Python/findUInitial.py
import numpy as np

from findKappa import findKappa
from findAlpha import findAlpha


def findUInitial(b, h, K, nu, eta, r, sigma, lambda_val, option):
    M = round(2*b/h)
    h = 2*b/M

    x = np.arange(-b, b+h, h)
    SbyK = np.exp(x)

    if(option == 'call'):
        v = []
        val_arr = K*SbyK-K
        for val in val_arr:
            v.append(max(val, 0))
    else:
        v = []
        val_arr = K-K*SbyK
        for val in val_arr:
            v.append(max(val, 0))

    kappa = findKappa(nu, eta)
    alpha = findAlpha(r, lambda_val, sigma, kappa)

    u = np.multiply(v, np.exp(-alpha*x))

    # Testing stuff
    # print(v)
    # print(np.exp(-alpha*x))
    # print(u)

    # for i in range(0, len(u)):
    #     u_val = u[i]
    #     if(u_val != v[i]*np.exp(-alpha*x)[i]):
    #         print(i, 'omg')
    #         break

    return u


# print(findUInitial(1.5, 3/128, 100, -0.9, 0.45, 0.05, 0.15, 0.1, 'call'))
